### ISSUE TITLE
Improving counterexamples for concurrent designs

### ISSUE THREAD

**Issue by craft095** [tlaplus/tlaplus#1084]
Let's consider a specification with three threads - A, B, and C - where each thread performs a series of sequential actions. You can get a counter-example that looks like:

```
A_Step0
B_Step0
C_Step0
A_Step1
B_Step1
C_Step1
A_Step2
B_Step2
C_Step2
```

I.e. all threads execute the first step, then all

It is often the case that such extensive interleaving is not essential for the problem found, and TLC could potentially generate a more comprehensible trace of the same length:

```
A_Step0
A_Step1
A_Step2
B_Step0
B_Step1
B_Step2
C_Step0
C_Step1
C_Step2

```

Such a transformation would be very helpful especially for long traces.

One solution is to ask the user to provide a function that calculates a "quality" metric for the trace (e.g., number of switches between thread). TLC could then iterate through counter-examples and report the one with the best metric.

Another approach is for TLC to choose the next state (during breadth-first exploration) in a specific order - based on the trace quality metric. This way it could generate better counterexamples without enumerating all of them.

**Comment by ahelwer** [comment]
For your first suggestion, I suppose this could be done by requiring TLC to finish checking a certain level (of steps from the initial state) before reporting an invariant violation. It would then feed all violations found into the user-provided ordering function and return the one with the highest score. Your second suggestion would require adding another branch to the hot path so I don't think it's a good idea.

The ordering function is a bit weird. It could require writing a predicate on pairs of steps. TLA+ does not really natively have the ability to express anything about pairs of steps, because they will share variables. I suppose it could take the form:
```
x ≺ y ≜ ...
```
where `x` is one step and `y` is another, then you could address the source and destination states of each step with `x`, `x'`, `y`, `y'`, and they are each a function from strings to variable values; so `x.varname` would provide the value of spec variable `varname` in the source state of step `x` and `y.varname'` would provide the value of spec variable `varname` in the destination state of step `y`. TLC would then output the trace of highest lexicographic order. If we don't want to use the prime operator (for example level-checking might not like this, I don't know) we could just use `x[1]` and x[2]` subscripts instead as each step is given as a 2-tuple of maps from strings to values.

Alternatively the ordering function could be defined on sequences of states. In this case `x` and `y` would be `Seq([STRING -> S])` where `S` is whatever values the variables can take on. This would let the user write more-powerful predicates over the entire trace, but predicates might be more difficult to write.

Anyway this seems like a substantial change to how TLC operators. I don't foresee anybody working on this soon, but if you would like to contribute dev time I suggest joining the [December TLA+ project meeting](https://groups.google.com/g/tlaplus/c/CpAEnrf-DHQ/m/YrORpIfSBwAJ) to talk it over with people.

**Comment by Calvin-L** [comment]
> Another approach is for TLC to choose the next state (during breadth-first exploration) in a specific order - based on the trace quality metric. This way it could generate better counterexamples without enumerating all of them.

I don't think this would be too difficult to implement. In fact, I did this exact thing back in 2021 (with some help from @lemmy). It was helpful for understanding a particular system where I wanted counterexamples with fewer context switches, even if those counterexamples were not the shortest possible.

The code I wrote for that project is clunky and has a lot of sharp edges, but here's an outline of what I did:

1. "a 'quality' metric for the trace" is too difficult; at runtime it is very expensive for TLC to reconstruct the entire trace. Instead, it is much easier to have the TLA+ spec track the number of context switches in a history variable so the "quality" can be read out of individual states.
2. I implemented a [StateQueue](https://github.com/tlaplus/tlaplus/blob/94a4bae0b4a217e53b1998980d3d6961c8fcf601/tlatools/org.lamport.tlatools/src/tlc2/tool/queue/StateQueue.java) backed by a thread-safe priority queue. The priority queue orders states by their "number of context switches" variable. In my case the queue was hard-coded to look for a particular variable name, so it was not general.
3. Hook up the new state queue [in `IStateQueue.get()`](https://github.com/tlaplus/tlaplus/blob/94a4bae0b4a217e53b1998980d3d6961c8fcf601/tlatools/org.lamport.tlatools/src/tlc2/tool/queue/IStateQueue.java#L31).

...and that's it! (Er, more or less, anyway.)
 
  - Big caveat: my queue was fully synchronized and therefore _very slow_. It also kept the entire state queue in memory, so it does not scale to large systems. It worked because my system was relatively small.

**Comment by ahelwer** [comment]
Does the TLC state graph store multiple predecessors or just the first encountered node that has a step to a given state?

**Comment by Calvin-L** [comment]
I'll have to double-check, but IIRC for the purposes of invariant/action violations, each state has exactly one predecessor. The predecessor is arbitrary ("the first encountered node that has a step to a given state"). The full state graph is also maintained for liveness checking and state graph dumping, but it is more difficult to access.

(Natural follow-up question: if every state has an arbitrary predecessor, how does my approach find traces with a minimal number of context switches? The history variable increases the state space. TLC will visit the same "logical state" multiple times, once for each number of possible context switches along traces that lead to that logical state. But, the one with the fewest context switches will be visited first.)

**Comment by ahelwer** [comment]
If each state has one predecessor then my initial comment is wrong and TLC won't be able to construct the optimal state trace when a violation is found; it will have to use a greedy ordering function over all next-state transitions in the hot path.

**Comment by lemmy** [comment]
FWIW: The following will make TLC exhaustively find *all* the shortest counterexamples when combined with the `-continue` flag. When using a single worker, this is deterministic. With multiple workers, it will be "sufficiently reliable".

```tla
---- MODULE Foo ----
...

YourInv ==
    ...

MinInv == 
    ~YourInv => ~TLCSet(1, Min({TLCGet("level"), TLCGetOrDefault(1, 9999)}))

MinConstraint ==
    TLCGet("level") < TLCGetOrDefault(1, 9999)

====

INVARIANT MinInv
CONSTRAINT MinConstraint
```

Assuming that `-dumpTrace` is compatible with `-continue`, a post-processing approach — similar to the one discussed in https://github.com/tlaplus/tlaplus/issues/400 — could be used to find the optimal counterexample.
